### PR TITLE
[NEW] Go to RoomView if the mentioned #channel is already joined

### LINKED ISSUE #4201: [NEW] When clicking on #referenced-channel, go to the channel view OR the channel info view
### Description:

When you click on a referenced channel (eg. #support), you will be directed to the info page of that channel. This was decided as a safe place to direct, as the user may or may not be on that channel already, or may or may not have permission to enter it.

It would be great to have the best of both: If the user is already on that channel, it will be directed to the channel directly. 
Otherwise, when the user isn't on that channel, it will be directed to the channel info page.

### Environment Information:

- Rocket.Chat Server Version: 4.7.0
- Rocket.Chat App Version:  4.28.0
- Device Name: Android
- OS Version: Xiaomi Latest

### Steps to reproduce:

1. make sure you are not on #channel-test
2. post a message anywhere with a reference to #channel-test
3. click on #channel-test
4. you should be presented with the channel info view
5. join the channel
6. go back to the message posted in step 2
7. click again on a reference to #channel-test

### Expected behavior:

You should be directed to the channel

### Actual behavior:

You are directed to the channel info view again, adding 1 unnecessary "tap" between you and the desired channel

### Additional context:

This will improve UX, as you are expecting to go directly to the channel, if you are already on that channel.

### PR DESCRIPTION
**Pull request by try-catch-stack**
## Issue(s)	

Fixes #4201 
## How to test or reproduce

* Mention a `#channel` that you haven't joined and click on it. You should be redirected to `RoomInfoView`  .
* Mention a `#channel` that you have already joined and click on it. You should be redirected to the `RoomView` directly.

## Screenshots

https://user-images.githubusercontent.com/77742477/169062333-ab52836a-797e-4bc4-b537-9e59de789923.mp4

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] Improvement (non-breaking change which improves a current function)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (if none of the other choices apply)

## Checklist

- [X] I have read the [CONTRIBUTING](https://github.com/RocketChat/Rocket.Chat/blob/develop/.github/CONTRIBUTING.md#contributing-to-rocketchat) doc
- [X] I have signed the [CLA](https://cla-assistant.io/RocketChat/Rocket.Chat.ReactNative)
- [X] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works (if applicable)
- [ ] I have added necessary documentation (if applicable)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by diegolmello — CHANGES_REQUESTED** [review]

File: app/containers/markdown/Hashtag.tsx [hunk 1]
```diff
@@ -15,15 +21,27 @@ interface IHashtag {
...
 				rid: channels?.[index]._id
 			};
-			navToRoomInfo(navParam);
+			const room = await getSubscriptionByRoomId(navParam.rid!);
+			if (room) {
+				if (isMasterDetail) {
+					// Close the modal if it is open and then redirect to the channel
+					Navigation.navigate('DrawerNavigator');
+					navigation.replace('RoomView', room);
+				} else navigation.push('RoomView', room);
```
Comment (diegolmello) [inline app/containers/markdown/Hashtag.tsx:40]:
```suggestion
				if (isMasterDetail) {
					// Close the modal if it is open and then redirect to the channel
					Navigation.navigate('DrawerNavigator');
					navigation.replace('RoomView', room);
				} else {
				    navigation.push('RoomView', room);
				}}
```

File: app/containers/markdown/Hashtag.tsx [hunk 2]
```diff
@@ -15,15 +21,27 @@ interface IHashtag {
...
+			} else {
+				// Navigate to RoomInfoView if the channel is not joined
+				navToRoomInfo(navParam);
```
Comment (diegolmello) [inline app/containers/markdown/Hashtag.tsx:43]: Did you notice on your video the title is getting blank on this step?
Also, try to interact a bit more on this, like joining, navigating to RoomActions, etc and make sure the layout is working correctly.

File: app/containers/markdown/Hashtag.tsx [hunk 3]
```diff
@@ -15,15 +21,27 @@ interface IHashtag {
...
-	const handlePress = () => {
+	const handlePress = async () => {
 		const index = channels?.findIndex(channel => channel.name === hashtag);
 		if (typeof index !== 'undefined' && navToRoomInfo) {
 			const navParam = {
 				t: 'c',
 				rid: channels?.[index]._id
 			};
-			navToRoomInfo(navParam);
+			const room = await getSubscriptionByRoomId(navParam.rid!);
```
Comment (diegolmello) [inline app/containers/markdown/Hashtag.tsx:34]: Try not to use non-null assertion (bang).
It's bad and might crash the app in production.
if necessary, do `if (navParam.rid) { ... }`

File: app/containers/markdown/Hashtag.tsx [hunk 4]
```diff
@@ -15,15 +21,27 @@ interface IHashtag {
...
+			if (room) {
+				if (isMasterDetail) {
+					// Close the modal if it is open and then redirect to the channel
+					Navigation.navigate('DrawerNavigator');
+					navigation.replace('RoomView', room);
```
Comment (diegolmello) [inline app/containers/markdown/Hashtag.tsx:39]: Maybe you should use `goRoom`.
It's going to be hard to maintain this in case we change `goRoom` logic.

**Review by diegolmello — CHANGES_REQUESTED** [review]

**Comment by GleidsonDaniel** [comment]
@try-catch-stack can you fix the tests?

**Comment by try-catch-stack** [comment]
Hey @GleidsonDaniel . I discussed this issue with @diegolmello on the channel. I tried few mocks for the libraries but didn't work so left it as it is for now. A point to be mentioned is, when I didn't use the `goRoom` method, the tests were fine.
Maybe some insights would help (:

**Comment by GleidsonDaniel** [comment]
Okay, I'll take a look and try to help you,

**Comment by GleidsonDaniel** [comment]
Done bro :)

**Comment by try-catch-stack** [comment]
Woah 🙇🏼 !! Thanks @GleidsonDaniel  :)

**Review by GleidsonDaniel — APPROVED** [review]

**Pull request merged**